# Hypothesis testing with eDNA metabarcoding

Data processing workflow and supplementary data for:

Harper *et al.* (2019) Testing ecological hypotheses at the pondscape with environmental DNA metabarcoding: a case study on a threatened amphibian

Permanently archived at: [![DOI](https://zenodo.org/badge/180420643.svg)](https://zenodo.org/badge/latestdoi/180420643)


## Background

eDNA samples were collected from 532 ponds across the UK as part of Natural England's Great Crested Newt Evidence Enhancement Programme. The samples were screened for great crested newt (*Triturus cristatus*) using targeted quantitative PCR (qPCR). Harper *et al*. (2018) reanalysed the samples using eDNA metabarcoding to compare the sensitivity of qPCR and metabarcoding for rare or low density species, such as the great crested newt. The dedicated GitHub repository for that study can be found [here](https://github.com/HullUni-bioinformatics/Harper_et_al_2018). The repository contains Jupyter notebooks used for bioinformatics analysis, R scripts used for data manipulation and data analysis, and associated data.

We repurposed the taxonomically assigned sequence reads generated by Harper *et al*. (2018) to investigate the potential of eDNA metabarcoding for ecological hypothesis testing at the pondscape, with the great crested newt as our focal species. We compare determinants of eDNA-based great crested newt occupancy to those reported in the existing literature, and test the status of the great crested newt as an umbrella species for pond conservation.


## Contents

R scripts used to analyse metabarcoding data and produce figures [(here)](https://github.com/lrharper1/Harper_et_al_2019_hypothesis_testing_with_eDNA_metabarcoding/tree/master/R%20scripts)

Data needed to run analyses in R [(here)](https://github.com/lrharper1/Harper_et_al_2019_hypothesis_testing_with_eDNA_metabarcoding/tree/master/Data/)


## Setting up the environment

In order to retrieve scripts and associated data, start by cloning this repository to your current directory:

```
git clone --recursive https://github.com/lrharper1/Harper_et_al_2019_hypothesis_testing_with_eDNA_metabarcoding.git
```

In order to make use of our scripts and data, you will have to install R on your computer. R is compatible with all major operating systems, but see CRAN (https://cran.r-project.org/) for details.

Once R is installed, open our R script in the R console and simply run the code!
